File: app/src/main/java/com/ShaHar91/poems/ui/home/categories/CategoryFragment.java
package com.shahar91.poems.ui.home.categories;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.shahar91.poems.R;
import com.shahar91.poems.data.models.Category;
import com.shahar91.poems.ui.base.BaseActivity;
import com.shahar91.poems.ui.base.normal.BaseGoogleFragment;
import com.shahar91.poems.ui.home.categories.adapter.CategoryAdapter;
import com.shahar91.poems.ui.home.poemsPerCategoryList.PoemsPerCategoryListFragment;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class CategoryFragment extends BaseGoogleFragment<CategoryViewModel, CategoryComponent> {
    private static final String TAG_POEMS_LIST = "TagPoemsList";

    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.rvCategories)
    RecyclerView rvCategories;

    private CategoryAdapter adapter;

    @Override
    protected CategoryComponent createComponent() {
        return DaggerCategoryComponent.builder()
                .applicationComponent(appComponent())
                .build();
    }

    public static CategoryFragment newInstance(boolean showBackIcon) {
        CategoryFragment fragment = new CategoryFragment();
        Bundle args = new Bundle();
        args.putBoolean(SHOW_BACK_ICON, showBackIcon);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        component.inject(this);
    }

    @Override
    public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_categories, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ButterKnife.bind(this, view);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        viewModel = new ViewModelProvider(this, viewModelFactory).get(CategoryViewModel.class);

        initViews();

        addDisposable(viewModel.getCategories().subscribe(this::showCategories, Throwable::printStackTrace));
    }

    private void showCategories(List<Category> categoryList) {
        adapter.setItems(categoryList);
    }

    private void initViews() {
        //toolbar
        toolbar.setTitle("Categories");
        configureToolbar(toolbar, null);

        adapter = new CategoryAdapter(getActivity(), this::handleClick);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getActivity());
        rvCategories.setLayoutManager(linearLayoutManager);

        rvCategories.setAdapter(adapter);
    }

    private void handleClick(Category category) {
        PoemsPerCategoryListFragment poemsPerCategoryListFragment = PoemsPerCategoryListFragment.newInstance(true, category);

        ((BaseActivity) requireActivity()).replaceFragment(R.id.flHomeContainer, poemsPerCategoryListFragment, TAG_POEMS_LIST, true);
    }

    @Override
    public void onResume() {
        super.onResume();

        viewModel.registerCategoryQuery();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        viewModel.stopListeningForChangesInBackend();
    }
}

File: app/src/main/java/com/ShaHar91/poems/ui/home/categories/CategoryFragment.kt
package com.shahar91.poems.ui.home.categories

import android.os.Bundle
import android.view.View
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import be.appwise.core.extensions.view.setupRecyclerView
import be.appwise.emptyRecyclerView.RecyclerViewState
import com.shahar91.poems.MyApp
import com.shahar91.poems.R
import com.shahar91.poems.databinding.FragmentCategoriesBinding
import com.shahar91.poems.ui.base.PoemBaseBindingVMFragment
import com.shahar91.poems.ui.home.categories.adapter.CategoryAdapter

class CategoryFragment : PoemBaseBindingVMFragment<FragmentCategoriesBinding>() {

    private val categoryAdapter = CategoryAdapter {
        CategoryFragmentDirections.actionCategoryFragmentToPoemsPerCategoryListFragment(it.id, it.name)
            .run(findNavController()::navigate)
    }

    override fun getLayout() = R.layout.fragment_categories
    override fun getToolbar() = mBinding.mergeToolbar.toolbar
    override val mViewModel: CategoryViewModel by viewModels { getViewModelFactory() }
    override fun getViewModelFactory() = CategoryViewModel.FACTORY(MyApp.categoryRepository)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        mBinding.run {
            viewModel = mViewModel.apply {
                getAllCategories()
            }
        }

        initViews()
        initObservers()
    }

    private fun initViews() {
        mBinding.run {
            rvCategories.run {
                setupRecyclerView(null)
                emptyStateView = emptyView
                loadingStateView = loadingView
                adapter = categoryAdapter
                state = RecyclerViewState.LOADING
            }

            themeSwipeToRefresh(srlRefreshCategories)
        }
    }

    private fun initObservers() {
        mViewModel.categoriesLive.observe(viewLifecycleOwner) {
            categoryAdapter.submitList(it)
            if (it.isNotEmpty()) {
                mBinding.rvCategories.state = RecyclerViewState.NORMAL
            }
        }
    }

    override fun onError(throwable: Throwable) {
        super.onError(throwable)
        mBinding.run {
            if (rvCategories.state == RecyclerViewState.LOADING) {
                rvCategories.state = RecyclerViewState.EMPTY
            }
            mViewModel.setIsRefreshing(false)
        }
    }
}